crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+					let _ = T::MultiCurrency::ensure_can_withdraw(&origin, Some(listing.payment_asset), amount, WithdrawReason::Reserve.into(), balance_after_bid)?;
+				}
+
+				// try lock funds
+				T::MultiCurrency::reserve(&origin, Some(listing.payment_asset), amount)?;
+
+				ListingWinningBid::<T>::mutate(&collection_id, token_id, |maybe_current_bid| {
+					if let Some(current_bid) = maybe_current_bid {
+						// replace old bid
+						T::MultiCurrency::unreserve(&current_bid.0, Some(listing.payment_asset), current_bid.1);
+					}
+					*maybe_current_bid = Some((origin, amount))
+				});
+
+				Self::deposit_event(RawEvent::Bid(collection_id, token_id, amount));
+			} else {
+				return Err(Error::<T>::NotForAuction.into());
+			}
+		}
+
+		/// Close a sale or auction
+		/// Requires no successful bids have been made for the auction.
+		/// Caller must be the token owner
+		#[weight = T::WeightInfo::cancel_sale()]
+		fn cancel_sale(origin, collection_id: CollectionId, token_id: T::TokenId) {
+			let origin = ensure_signed(origin)?;
+			let current_owner = Self::token_owner(&collection_id, token_id);
+			ensure!(current_owner == origin, Error::<T>::NoPermission);
+
+			match Self::listings(&collection_id, token_id) {
+				Some(Listing::<T>::DirectSale(sale)) => {
+					ListingEndSchedule::<T>::mutate(sale.close, |schedule| schedule.retain(|(c, t)| (c, t) != (&collection_id, &token_id)));
+					Listings::<T>::remove(&collection_id, token_id);
+					Self::deposit_event(RawEvent::DirectSaleClosed(collection_id, token_id));
+				},
+				Some(Listing::<T>::Auction(auction)) => {
+					ensure!(Self::listing_winning_bid(&collection_id, token_id).is_none(), Error::<T>::TokenListingProtection);
+					Listings::<T>::remove(&collection_id, token_id);
+					ListingEndSchedule::<T>::mutate(auction.close, |schedule| schedule.retain(|(c, t)| (c, t) != (&collection_id, &token_id)));
+					Self::deposit_event(RawEvent::AuctionClosed(collection_id, token_id, AuctionClosureReason::VendorCancelled));
+				},
+				None => {},
+			}
+		}
+
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Transfer ownership of a token. modifies storage, does no verification, infallible.
+	fn transfer_ownership(
+		collection_id: &[u8],
+		token_id: T::TokenId,
+		current_owner: &T::AccountId,
+		new_owner: &T::AccountId,
+	) {
+		// Update token ownership
+		<CollectedTokens<T>>::mutate(collection_id, current_owner, |tokens| tokens.retain(|t| t != &token_id));
+		<TokenOwner<T>>::insert(collection_id, token_id, new_owner);
+		<CollectedTokens<T>>::append(collection_id, new_owner, token_id);
+	}
+	/// Remove a single direct listing and all it's metadata
+	fn remove_direct_listing(collection_id: &CollectionId, token_id: T::TokenId) {
+		let listing_type = Listings::<T>::take(collection_id, token_id);
+		ListingWinningBid::<T>::remove(collection_id, token_id);
+		if let Some(Listing::<T>::DirectSale(listing)) = listing_type {
+			ListingEndSchedule::<T>::mutate(listing.close, |listings| {
```

Comment:
With BtreeSet instead of Vec for tha values of this map, you can simply use the `remove` function here.